### PR TITLE
[Merged by Bors] - feat(RingTheory/Flat): flat is a local property of ring homomorphisms

### PR DESCRIPTION
**Pull request by chrisflav**
We show that `RingHom.Flat` satisfies `RingHom.PropertyIsLocal`.

---

- [x] depends on: #19786
- [x] depends on: #19788

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/RingHom/Flat.lean [hunk 1]
```diff
@@ -68,4 +68,29 @@ lemma holdsForLocalizationAway : HoldsForLocalizationAway Flat := by
...
+  algebraize_only [f]
+  replace h : ∀ r : s, Module.Flat R (Localization.Away (r : S)) := by
+    intro r; simp_rw [RingHom.Flat] at h;
+    convert h r; ext a x;
+    have : SMul.smul a x = (algebraMap R (Localization.Away r.val)) a * x := by
+      rw [← Algebra.smul_def]
+      rfl
+    rw [this]
+    rfl
+  let g (r : s) := (Algebra.linearMap S (Localization.Away r.val))
```
Comment (joelriou) [inline Mathlib/RingTheory/RingHom/Flat.lean:82]:
```suggestion
  let g (r : s) := Algebra.linearMap S (Localization.Away r.val)
```

File: Mathlib/RingTheory/RingHom/Flat.lean [hunk 2]
```diff
@@ -68,4 +68,29 @@ lemma holdsForLocalizationAway : HoldsForLocalizationAway Flat := by
     rw [RingHom.Flat]; convert this; ext; simp_rw [Algebra.smul_def]; rfl
   exact IsLocalization.flat _ (Submonoid.powers r)
 
+lemma ofLocalizationSpanTarget : OfLocalizationSpanTarget Flat := by
+  introv R hsp h
+  algebraize_only [f]
+  replace h : ∀ r : s, Module.Flat R (Localization.Away (r : S)) := by
+    intro r; simp_rw [RingHom.Flat] at h;
+    convert h r; ext a x;
```
Comment (joelriou) [inline Mathlib/RingTheory/RingHom/Flat.lean:76]:
```suggestion
    intro r; simp_rw [RingHom.Flat] at h
    convert h r; ext a x
```

File: Mathlib/RingTheory/RingHom/Flat.lean [hunk 3]
```diff
@@ -68,4 +68,29 @@ lemma holdsForLocalizationAway : HoldsForLocalizationAway Flat := by
...
+    have : SMul.smul a x = (algebraMap R (Localization.Away r.val)) a * x := by
```
Comment (joelriou) [inline Mathlib/RingTheory/RingHom/Flat.lean:77]:
```suggestion
    have : SMul.smul a x = algebraMap R (Localization.Away r.val) a * x := by
```

File: Mathlib/RingTheory/RingHom/Flat.lean [hunk 4]
```diff
@@ -68,4 +68,29 @@ lemma holdsForLocalizationAway : HoldsForLocalizationAway Flat := by
...
+      rw [← Algebra.smul_def]
+      rfl
+    rw [this]
+    rfl
+  let g (r : s) := (Algebra.linearMap S (Localization.Away r.val))
+  exact Module.flat_of_isLocalized_span _ _ s hsp _ g h
```
Comment (alreadydone) [inline Mathlib/RingTheory/RingHom/Flat.lean:83]: This works:
```suggestion
  refine Module.flat_of_isLocalized_span _ _ s hsp _
    (fun r ↦ Algebra.linearMap S <| Localization.Away r.1) ?_
  dsimp only [RingHom.Flat] at h
  convert h; ext
  apply Algebra.smul_def
```

**Review by alreadydone — APPROVED** [review]

**Comment by alreadydone** [comment]
maintainer delegate

**Comment by riccardobrasca** [comment]
Thanks!

bors d+

**Comment by chrisflav** [comment]
Thanks for the reviews!

bors r+

**Pull request closed without merging**